File: src/main/java/me/itzisonn_/meazy/parser/ast/expression/literal/NullLiteral.java
package me.itzisonn_.meazy.parser.ast.expression.literal;

import me.itzisonn_.meazy.instruction.InstructionsSet;
import me.itzisonn_.meazy.parser.DataType;
import me.itzisonn_.meazy.parser.ast.ProgramUnit;
import me.itzisonn_.meazy.parser.ast.expression.Expression;
import me.itzisonn_.meazy.runtime.environment.Environment;
import org.jspecify.annotations.NullMarked;

import java.lang.constant.ConstantDescs;

@NullMarked
public class NullLiteral implements Expression {
    @Override
    public void emit(InstructionsSet instructionsSet, Environment environment, ProgramUnit parent) {
        instructionsSet.loadNull();
    }

    @Override
    public DataType getType(Environment environment, ProgramUnit parent) {
        return DataType.ofNullable(ConstantDescs.CD_Object);
    }
}

File: src/main/kotlin/me/itzisonn_/meazy/parser/ast/expression/literal/NullLiteral.kt
package me.itzisonn_.meazy.parser.ast.expression.literal

import me.itzisonn_.meazy.instruction.InstructionsSet
import me.itzisonn_.meazy.parser.DataType
import me.itzisonn_.meazy.parser.ast.ProgramUnit
import me.itzisonn_.meazy.parser.ast.expression.Expression
import me.itzisonn_.meazy.runtime.environment.Environment
import java.lang.constant.ConstantDescs

class NullLiteral : Expression {
    override fun emit(instructions: InstructionsSet, environment: Environment, parent: ProgramUnit) {
        instructions.loadNull()
    }

    override fun getType(environment: Environment, parent: ProgramUnit): DataType {
        return DataType.ofNullable(ConstantDescs.CD_Object)
    }
}
